core/src/main/scala/sqlitis/Parser.scala

```diff
@@ -46,83 +71,92 @@ object Parser {
+        expr.map[Field[Unit]](ExpressionField(_, None))
+    )
+
+  private lazy val from =
+    token[From](
+      (identifier ~ whitespace ~ identifier).map { case ((t, _), a) => TableName(t.name, Some(a.name)) } |
+        identifier.map(i => TableName(i.name, None))
+    )
+
+  private lazy val orderBy = (
+    expr ~ (kw("ASC").as(true) | kw("DESC").as(false)).backtrack.?.map(_.getOrElse(false))
+  ).map { case (e, asc) => OrderBy[Unit](e, asc) }
+
+  private lazy val int: Parser1[Int] = Numbers.signedIntString.map(_.toInt)
+
+  implicit lazy val expr: Parser1[Expression[Unit]] =
+    defer1(
+      opsByPrecedence
+        .foldLeft[Parser1[Expression[Unit]]](singleExpr) { case (e, level) =>
+          val op: Parser1[Op] = (
+            P.oneOf1(level.ops.toList.map(o => (string1(o.sym)).as(o))) <* whitespace
+          ).backtrack
+
+          if (level.leftAssociative)
+            (e ~ (op ~ e).rep).map { case (i, t) =>
               t.foldLeft(i) { case (a, (o, b)) => o.to(a, b) }
-          }
-        else
-          (many(e ~ op) ~ e).map {
-            case (t, i) =>
+            }
+          else
+            ((e ~ op).backtrack.rep.with1 ~ e).map { case (t, i) =>
               t.foldRight(i) { case ((a, o), b) => o.to(a, b) }
-          }
-
-    }
-    .named("expr")
+            }
 
-  private val field: Parser[Field[Unit]] = token(
-    (expr ~ kw("AS") ~ identifier).map[Field[Unit]] { case ((e, _), i) => ExpressionField(e, Some(i.name)) } |
-      (identifier <~ kw(".") <~ kw("*")).map[Field[Unit]](i => Splat(Some(i.name))) |
-      kw("*").map[Field[Unit]](_ => Splat(None)) |
-      expr.map[Field[Unit]](ExpressionField(_, None))
-  )
-
-  private val from = token[From](
-    (identifier ~ whitespace ~ identifier).map { case ((t, _), a) => TableName(t.name, Some(a.name)) } |
-      identifier.map(i => TableName(i.name, None))
-  )
+        }
+    )
 
-  private val orderBy = (
-    expr ~ opt(kw("ASC").map(_ => true) | kw("DESC").map(_ => false)).map(_.getOrElse(false))
-  ).map((OrderBy[Unit](_, _)).tupled)
-
-  implicit val select: Parser[Select[Unit]] = (
-    kw("SELECT") ~> opt(kw("DISTINCT")).map(_.isDefined) ~
-      sepBy(field, comma) ~
-      optKw("FROM", sepBy(from, comma)).map(_.getOrElse(Nil)) ~
-      optKw("WHERE", expr) ~
+  implicit lazy val select: P[Select[Unit]] = (
+    kw("SELECT") *> (kw("DISTINCT")).?.map(_.isDefined) ~
+      repSep(field, 0, comma) ~
+      optKw("FROM", repSep(from, 1, comma)).map(_.getOrElse(Nil)) ~
+      optKw("WHERE", expr <* maybeWhitespace) ~
       optKw("ORDER BY", orderBy) ~
       optKw("LIMIT", token(int))
-  ).map {
-    case (((((isDistinct, fields), from), where), orderBy), limit) =>
-      Select(
-        isDistinct = isDistinct,
-        fields = fields,
-        from = from,
-        where = where,
-        orderBy = orderBy,
-        limit = limit
-      )
+  ).map { case (((((isDistinct, fields), from), where), orderBy), limit) =>
```

Comment:
I'd use cats syntax here and do, `(p1, p2, p3, ...).mapN { (a1, a2, ...`

and not have to deal with the nesting personally.  But this is only a syntactic change.